File: reference/springboot/adapters-rest/src/main/java/org/example/adapters/rest/AnimalsController.java
package org.example.adapters.rest;

import org.example.adapters.rest.dto.AnimalDto;
import org.example.adapters.rest.dto.GetAnimalResponse;
import org.example.adapters.rest.dto.GetAnimalsResponse;
import org.example.domain.usecases.CreateAnimalUseCase;
import org.example.domain.usecases.GetAnimalUseCase;
import org.example.domain.usecases.GetAnimalsUseCase;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.util.UriComponentsBuilder;

import java.util.UUID;

@RestController
public class AnimalsController extends AbstractAnimalsApi {

    public AnimalsController(GetAnimalUseCase getAnimalUseCase, GetAnimalsUseCase getAnimalsUseCase, CreateAnimalUseCase createAnimalUseCase) {
        super(getAnimalUseCase, getAnimalsUseCase, createAnimalUseCase);
    }

    @Override
    public ResponseEntity<GetAnimalsResponse> getAnimals(int limit, int offset) {
        var response = super.getAnimals(limit, offset);

        // discrepancy, path duplication. Is there a possibility to construct url bases on an operation?
        var nextLink = UriComponentsBuilder.fromPath("/animals")
            .queryParam("limit", limit)
            .queryParam("offset", offset + limit)
            .build()
            .toUri();

        response.getBody().getLinks().setNext(nextLink);
        return response;
    }

    @Override
    public ResponseEntity<GetAnimalResponse> getAnimal(UUID animal_id) {
        return super.getAnimal(animal_id);
    }

    @Override
    public ResponseEntity<Void> createAnimal(AnimalDto animal) {
        super.createAnimal(animal);
         // discrepancy, path duplication. Is there a possibility to construct url bases on an operation?
        var location = UriComponentsBuilder.fromPath("/animals/{animal_id}").build(animal.getAnimalId());
        return ResponseEntity.created(location).build();
    }
}

File: reference/dotnet/Adapters/Company.Product.Adapters.Rest/Controllers/AnimalsController.cs
namespace Company.Product.Adapters.Rest.Controllers;

public class AnimalsController : AbstractAnimalsController1
{
    public AnimalsController(ICreateAnimalUseCase createAnimalUseCase, IGetAnimalUseCase getAnimalUseCase, IGetAnimalsUseCase getAnimalsUseCase) 
        : base(createAnimalUseCase, getAnimalUseCase, getAnimalsUseCase)
    { }

    public override async Task<Results<ForbidHttpResult, UnauthorizedHttpResult, Created>> CreateAnimal([FromBody, Required] Animal animal, CancellationToken cancellationToken)
    {
        await base.CreateAnimal(animal, cancellationToken);
        var location = Url.Action(
            action: nameof(GetAnimal),
            values: new { animal.AnimalId }
        );
        return TypedResults.Created(location);
    }

    public override async Task<Results<ForbidHttpResult, UnauthorizedHttpResult, NotFound, Ok<GetAnimalsResponse>>> GetAnimals([FromQuery, Required] int limit, [FromQuery, Required] int offset, CancellationToken cancellationToken)
    {
        var results = await base.GetAnimals(limit: limit, offset: offset, cancellationToken: cancellationToken);

        var ok = (Ok<GetAnimalsResponse>)results.Result;
        
        ok.Value.Links = new OffsetResponseLinks()
        {
            Next = Url.Action(
                action: nameof(GetAnimals),
                values: new { limit, offset = offset + limit }
            )
        };

        return results;
    }

    public override Task<Results<ForbidHttpResult, UnauthorizedHttpResult, NotFound, Ok<GetAnimalResponse>>> GetAnimal([FromRoute, Required] Guid animalId, CancellationToken cancellationToken)
    {
        return base.GetAnimal(animalId, cancellationToken);
    }
}